scripts/tests/sanitycheck/test_reporting_testsuite.py

```diff
@@ -0,0 +1,168 @@
+    Test 1: Check if apply_filters function has been run before running
+    discard_report
+    Test 2: Test if the generated report is not empty
+    Test 3: Test if the gerenrated report contains the expected columns"""
+    class_testsuite.platforms = platforms_list
+    class_testsuite.testcases = all_testcases_dict
+    filename = tmpdir.mkdir("test_discard").join("discard_report.csv")
+    with pytest.raises(SystemExit):
+        class_testsuite.discard_report(filename)
+    assert "apply_filters() hasn't been run!" in caplog.text
+
+    kwargs = {"exclude_tag" : ['test_a'], "exclude_platform" : ['demo_board_1'],
+              "platform" : ['demo_board_2']}
+    class_testsuite.apply_filters(**kwargs)
+    class_testsuite.discard_report(filename)
+    assert os.stat(filename).st_size != 0
+    with open(filename, "r") as file:
+        csv_reader = csv.reader(file)
+        assert set(['test', 'arch', 'platform', 'reason']) == set(list(csv_reader)[0])
+
+def test_csv_report(class_testsuite, instances_fixture, tmpdir):
+    """ Testing csv_report function of Testsuite class in sanitycheck
+    Test 1: Assert the csv_report isnt empty after execution of csv_report function
+    Test 2: Assert on the columns and values of the generated csv_report"""
+    class_testsuite.instances = instances_fixture
+    filename = tmpdir.mkdir("test_csv").join("sanitycheck_csv_report.csv")
+    class_testsuite.csv_report(filename)
+    assert os.path.exists(filename)
+    assert os.stat(filename).st_size != 0
+
+    mydict = {'test': [], 'arch' : [], 'platform' : [], 'status': [],
+              'extra_args': [], 'handler': [], 'handler_time': [],
+              'ram_size': [], 'rom_size': []}
+
+    with open(filename, "r") as file:
+        csv_reader = csv.reader(file)
+        assert set(mydict.keys()) == set(list(csv_reader)[0])
+
+    for instance in class_testsuite.instances.values():
+        mydict["test"].append(instance.testcase.name)
+        mydict["arch"].append(instance.platform.arch)
+        mydict["platform"].append(instance.platform.name)
+        instance_status = instance.status if instance.status is not None else ""
+        mydict["status"].append(instance_status)
+        args = " ".join(instance.testcase.extra_args)
+        mydict["extra_args"].append(args)
+        mydict["handler"].append(instance.platform.simulation)
+        mydict["handler_time"].append(instance.metrics.get("handler_time", ""))
+        mydict["ram_size"].append(instance.metrics.get("ram_size", '0'))
+        mydict["rom_size"].append(instance.metrics.get("rom_size", '0'))
+
+    dict_file = open(filename, "r")
+    dict_reader = csv.DictReader(dict_file)
+    columns = defaultdict(list)
+    for row in dict_reader:
+        for (key, value) in row.items():
+            columns[key].append(value)
+    for i in range(len(mydict)):
+        assert set(list(columns.values())[i]) == set(list(mydict.values())[i])
```

Comment:
why the cast to set() ? also, couldn't you do something like
```
for i, v in enumerate(mydict):
	assert list(columns.values())[i] == mydict[v]
```

even better, if columns have the same key you could do:
```
for i, v in enumerate(mydict):
	assert columns[v] == mydict[v]
```

couldn't ?

Another thing, `mydict` is a dictionary but `columns` is a `collections.defaultdict` why ?

Reply:
btw, I would use the same default dictionary since you have few elements

Reply:
Yes I missed that. Changed the  content in for loop to "assert columns[v] == mydict[v]".

I am using defaultdict(list) here as when the list class is passed as the default_factory argument, then a defaultdict is created with the values that are list. Since I am initiating the columns dictionary and appending values by doing "columns[key].append(value)" , the initial values of column[key] dictionary is empty which raises the keyerror. Else, I would have to initialize a list() for each entry in the dictionary. And using defaultdict  never raises the keyerror.

Reply:
> btw, I would use the same default dictionary since you have few elements.

I didnt get this part. "mydict" is for the expected values and columns dictionary is for reading the values that exist in the csv report.

Reply:
why are you using a different type of dict `collections.defaultdict` ?

Reply:
In this use case I am grouping the dictionary values according to the column headers like "test", "arch", platform" etc. All these are keys in the defaultdict. And the values of each one is the list grouped together according to keys. Now I am using defaultdict(list) here to initialize the keys before first use, or else we would get KeyError (if using normal mutable dictionary). Default dictionary gives two benefits: 1. Doesnt give keyError if the key doesnt exist. 2. If there is a missing key, then it calls list() to create new empty list, inserts the empty list into dictionary using missing key as key and then returns the reference to the list. It is basically providing an easier means to group the items (dictionary values in this case.).